Sources_v3/StreamChatUI/ChatChannel/ChatMessageListVC.swift

```diff
@@ -0,0 +1,333 @@
+        })
+
+        actions.append(.copy { [weak self] in
+            UIPasteboard.general.string = message.text
+            self?.dismiss(animated: true)
+        })
+
+        if message.isSentByCurrentUser {
+            actions.append(.edit { [weak self] in
+                debugPrint("edit")
+                self?.dismiss(animated: true)
+            })
+            actions.append(.delete { [weak self] in
+                self?.router.showMessageDeletionConfirmationAlert { confirmed in
+                    guard confirmed else { return }
+
+                    messageController.deleteMessage { [messageController] _ in
+                        self?.dismiss(animated: true)
+                        _ = messageController
+                    }
+                }
+            })
+        } else {
+            if currentUser.mutedUsers.contains(message.author) {
+                actions.append(.unmuteUser { [weak self] in
+                    let userController = messageController.client.userController(userId: message.author.id)
+                    userController.unmute { [userController] _ in
+                        self?.dismiss(animated: true)
+                        _ = userController
+                    }
+                })
+            } else {
+                actions.append(.muteUser { [weak self] in
+                    let userController = messageController.client.userController(userId: message.author.id)
+                    userController.mute { [userController] _ in
+                        self?.dismiss(animated: true)
+                        _ = userController
+                    }
+                })
+            }
+        }
+
+        return actions
+    }
+}
+
+public extension ChatMessageListVC.DataSource {
+    static func wrap<T: ChatMessageListVCDataSource>(_ ds: T) -> ChatMessageListVC.DataSource where T.ExtraData == ExtraData {
+        ChatMessageListVC.DataSource(
+            numberOfMessages: { [unowned ds] in ds.numberOfMessagesInChatMessageListVC($0) },
+            messageAtIndex: { [unowned ds] in ds.chatMessageListVC($0, messageAt: $1) },
+            loadMoreMessages: { [unowned ds] in ds.loadMoreMessagesForChatMessageListVC($0) },
+            replyMessageForMessageAtIndex: { [unowned ds] in ds.chatMessageListVC($0, replyMessageFor: $1, at: $2) },
+            controllerForMessage: { [unowned ds] in ds.chatMessageListVC($0, controllerFor: $1) }
+        )
+    }
+
+    static func empty() -> ChatMessageListVC.DataSource {
+        ChatMessageListVC.DataSource(
+            numberOfMessages: { _ in 0 },
+            messageAtIndex: { _, _ in
+                fatalError("Method shouldn't be called on empty data source")
+            },
+            loadMoreMessages: { _ in },
+            replyMessageForMessageAtIndex: { _, _, _ in nil },
+            controllerForMessage: { _, _ in
+                fatalError("Method shouldn't be called on empty data source")
+            }
+        )
+    }
```

Comment:
What was the reasoning for having `empty` instead of making the data source optional?

Reply:
Whole controller doesn't make much sense without datasource. So datasource is required.
Since we cannot use custom init, it needs to be either:
- optional: need to unwrap it everywhere and fallback to something, makes it dirty and just pollute codebase
- implicitly unwrapped optional: better "fall early" approach
- empty: middle ground that plays safe, but solves optional pollution